domain/src/main/java/com/github/ianellis/shifts/domain/di/shifts/ShiftsDomainModule.kt

```diff
@@ -37,32 +38,32 @@ class ShiftsDomainModule {
         locationRepository: LocationRespoitory,
         shiftRepository: ShiftRepository,
         clock: Clock
-    ): () -> @JvmSuppressWildcards Deferred<Unit> {
-        return EndShiftCommand(locationRepository, shiftRepository, clock, Dispatchers.IO)
+    ): Action {
+        return EndShiftCommand(locationRepository, shiftRepository, clock)
     }
 
     @Provides
     @LoadShifts
     fun providesLoadShifts(
         repository: ShiftRepository
-    ): () -> Unit {
+    ): Action {
         return LoadShiftsCommand(repository)
     }
 
     @Provides
     fun providesListenForShifts(
         shiftRepository: ShiftRepository,
         clock: Clock
-    ): Function1<@JvmSuppressWildcards Function1<@JvmSuppressWildcards Event<List<Shift>>, kotlin.Unit>, @JvmSuppressWildcards Function0<kotlin.Unit>> {
-        return ListenForShiftUpdatesCommand(shiftRepository, clock)
+    ): ReceiveChannel<Event<List<Shift>>> {
+        return ListenForShiftUpdatesCommand(shiftRepository, clock)()
     }
 
     @Provides
     @ActivityScope
     fun providesShiftRepository(
         service: ShiftsService
     ): ShiftRepository {
-        return ShiftRepository(service, ShiftsMemoryCache(), Dispatchers.IO, LatestEventEmitter())
+        return ShiftRepository(service, ShiftsMemoryCache())
```

Comment:
Does it make sense to make repository a singleton? The internal cache of the repository can be shared across the whole app.

Reply:
The Repo is not a global singleton, but a scoped singleton (guessing you got this). It needs to be, so that all of its clients are referring to the same instance, allowing it to manage it state effectively. 

Generally I would gate the access to the cache layer through a repository - this allows us to encapsulate our data access rule eg network with cache fallback. In a larger application you might want to share a single cache instance across multiple repositories - although I would generally try to avoid this